### PR TITLE
Initial typing of a serialized analysis

### PR DESCRIPTION
**Pull request by rictic**
* Stubbed out implementation of the `Context` object, which I've named `Analysis` (it's what you get from asking an analyzer to analyze!
* Wrote out and documented interfaces for the metadata format, which I've called `SerializedAnalysis`.
  * Wrote out an example for paper-button here: https://gist.github.com/rictic/b6fd4f19b6f265753d5d11cdb8a41940

### CODE REVIEW

File: src/analysis.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import {DocumentDescriptor} from './ast/ast';
+import {SerializedAnalysis} from './serialized-analysis';
+
+export class Analysis {
```
Comment (justinfagnani) [inline src/analysis.ts]: I was thinking `AnalysisResult`

File: src/analysis.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import {DocumentDescriptor} from './ast/ast';
```
Comment (justinfagnani) [inline src/analysis.ts]: license

File: src/analysis.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import {SerializedAnalysis} from './serialized-analysis';
+
+export class Analysis {
+  private descriptors_: DocumentDescriptor[];
```
Comment (justinfagnani) [inline src/analysis.ts]: Use leading underscore

File: src/analysis.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  constructor(descriptors: DocumentDescriptor[]) {
+    this.descriptors_ = descriptors;
+  }
+
+  serialize(): SerializedAnalysis {
```
Comment (justinfagnani) [inline src/analysis.ts]: Ultimately, I think this should be a function that takes an Analysis, not a method.

File: src/analyzer.ts [hunk 5]
```diff
@@ -91,6 +92,11 @@ export class Analyzer {
     return promise;
   }
 
+  async resolve(): Promise<Analysis> {
+    const descriptorPromises = Array.from(this._documentDescriptors.values());
+    return new Analysis(Array.from(await Promise.all(descriptorPromises)));
```
Comment (justinfagnani) [inline src/analyzer.ts]: `await Promise.all(...)` is already an `Array`

File: src/analyzer.ts [hunk 6]
```diff
@@ -91,6 +92,11 @@ export class Analyzer {
     return promise;
   }
 
+  async resolve(): Promise<Analysis> {
+    const descriptorPromises = Array.from(this._documentDescriptors.values());
```
Comment (justinfagnani) [inline src/analyzer.ts]: `Promise.all()` takes an iterable, no need for `Array.from()`

File: src/serialized-analysis.ts [hunk 7]
```diff
@@ -0,0 +1,425 @@
+export interface SerializedAnalysis { packages: Package[]; }
+
+export interface Package {
```
Comment (justinfagnani) [inline src/serialized-analysis.ts]: I don't think we need `Package`. An `elements.json` file should already be in a package, next to a `bower.json` and/or `package.json`. No need to duplicate that info here.

File: src/serialized-analysis.ts [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+   *
+   * e.g. `paper-input.html` or `app-toolbar/app-toolbar.html` (given that
+   * app-toolbar lives in the app-layout package).
+   */
+  path: string;
+
+  /** The tagname that the element registers itself as. e.g. `paper-input` */
+  tagname: string;
+
+  /** A markdown description for the element. */
```
Comment (justinfagnani) [inline src/serialized-analysis.ts]: We need to consider what to do if we read the documentation from an HTML comment...

File: src/serialized-analysis.ts [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+  description: string;
+
+  /**
+   * Paths, relative to `this.path` to demo pages for the element.
```
Comment (justinfagnani) [inline src/serialized-analysis.ts]: Not sure they should be relative to `this.path`: what about `/elements/foo-element.html` and `demos/foo-element-demo.html`? I think for simplicity all paths should be relative to this file.

File: src/serialized-analysis.ts [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+  attributes?: Attribute[];
+
+  /** The properties that this element has. */
+  properties?: Property[];
+
+  /** The events that this element fires. */
+  events?: Event[];
+
+  /** The shadow dom content slots that this element accepts. */
+  'slots': Slot[];
```
Comment (justinfagnani) [inline src/serialized-analysis.ts]: remove quotes

File: src/serialized-analysis.ts [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+
+  /** Information useful for styling the element and its children. */
+  styling: {
+
+    /** CSS Classes that produce built-in custom styling for the element. */
+    classes: {
```
Comment (justinfagnani) [inline src/serialized-analysis.ts]: I'm not sure this should just be classes. An element could have a `:host([height=5])` rule. Maybe just selectors?

File: src/serialized-analysis.ts [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+
+  /**
+   * The type that the attribute will be serialized/deserialized as.
+   *
+   * e.g. `string`, `number`, `boolean`, `RegExp`, `Array`, `Object`.
+   */
+  type?: string;
+
+  /** The default value of the attribute, if any. */
+  defaultValue?: string;
```
Comment (justinfagnani) [inline src/serialized-analysis.ts]: We need some way to indicate if an attribute is associated with a property... TBD?

File: src/serialized-analysis.ts [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+    // display: block or inline or whatever.
+  };
+
+  /**
+   * An extension point for framework-specific metadata, as well as any
+   * metadata not yet standardized here such as what polyfills are needed,
+   * behaviors and mixins used, the framework that the element was written in,
+   * tags/categories, links to specs that the element implements, etc.
+   */
+  metadata: Object;
```
Comment (justinfagnani) [inline src/serialized-analysis.ts]: How about we define a super-interface with `metadata` and `location`?

File: src/serialized-analysis.ts [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+  /**
+   * An extension point for framework-specific metadata, as well as any
+   * metadata not yet standardized here such as associated events and
+   * observers.
+   */
+  metadata?: Object;
+}
+
+export interface Event {
+  /** The name of the event. */
```
Comment (justinfagnani) [inline src/serialized-analysis.ts]: We also need some way to associate a change event with a property or attribute...

File: src/analysis.ts (hunk 2 above)
Comment (rictic) [inline src/analysis.ts]: Done.

File: src/analysis.ts (hunk 1 above)
Comment (rictic) [inline src/analysis.ts]: Seems unnecessarily wordy. The result of the analyzer is naturally an analysis.

OTOH `breakfast food` is an actual phrase people use, so, I guess there's precedent.

File: src/analyzer.ts (hunk 6 above)
Comment (rictic) [inline src/analyzer.ts]: Oh, neat!

File: src/analyzer.ts (hunk 5 above)
Comment (rictic) [inline src/analyzer.ts]: I can't imagine what I was thinking when I wrote that.

File: src/serialized-analysis.ts (hunk 8 above)
Comment (rictic) [inline src/serialized-analysis.ts]: Most of the elements we maintain are documented in HTML comments, but they still use markdown. e.g. https://github.com/PolymerElements/paper-button/blob/master/paper-button.html

File: src/serialized-analysis.ts (hunk 7 above)
Comment (rictic) [inline src/serialized-analysis.ts]: `elements.json`, I like it.

`version` was requested from the catalog folks, which prompted the concept of a package.

I'm ok with not duplicating info from bower.json and package.json and dropping those fields, I kinda threw them in on a lark. Do we ever want to represent multiple packages' worth of elements in one metadata file though? I suspect that we do for the catalog use case, where it would be nice to have one file that contained all of the metadata. For an `elements.json` it would make sense to serialize just one package, sure.

File: src/serialized-analysis.ts (hunk 9 above)
Comment (rictic) [inline src/serialized-analysis.ts]: I'm not sure I understand, can you rephrase?

File: src/serialized-analysis.ts (hunk 10 above)
Comment (rictic) [inline src/serialized-analysis.ts]: Done. clang-format gets a little confused and doesn't do the best thing, but leaving the quotes in confuses the json schema generator I'm using even more.

(spoiler: I've got json schema generation working. Also vscode understands json schema, and it's pretty great)

File: src/serialized-analysis.ts (hunk 14 above)
Comment (rictic) [inline src/serialized-analysis.ts]: Added a comment.

Is that needed though? What's the use case there?

File: src/analysis.ts (hunk 4 above)
Comment (rictic) [inline src/analysis.ts]: Acknowledged, but why's that?

File: src/analysis.ts (hunk 3 above)
Comment (rictic) [inline src/analysis.ts]: Done

File: src/serialized-analysis.ts (hunk 12 above)
Comment (rictic) [inline src/serialized-analysis.ts]: Added TBD comment

File: src/serialized-analysis.ts (hunk 13 above)
Comment (rictic) [inline src/serialized-analysis.ts]: Done

File: src/serialized-analysis.ts (hunk 9 above)
Comment (justinfagnani) [inline src/serialized-analysis.ts]: rather that "this file" (by which I meant the elements.json file) we can say package directory

**Comment by justinfagnani** [comment]
LGTM
